### PR TITLE
feat(async): add jitter to retry exponential backoff

### PR DESCRIPTION
**Pull request by lionel-rowe**
Fixes https://github.com/denoland/deno_std/issues/3378

### CODE REVIEW

**Comment by lino-levan** [comment]
I'm pretty strongly -1 against complicating the API to allow for arbitrary backoff functions. When I made the PR to introduce `async/retry`, I made a conscious decision to not do that (because outside of exponential backoff, what algorithm would you need?). I chose not to add jitter at the time because I didn't understand what the use case was but I think you did a good job in #3378 explaining why it might be useful.

TLDR; I think adding an option for jitter would be fine, but please don't overcomplicate the API with backoff functions.

**Comment by lino-levan** [comment]
I never thought about making jitter enabled by default but honestly that makes sense (plus it removes an option which is always nice). Thoughts @iuioiua?

**Comment by lionel-rowe** [comment]
> I never thought about making jitter enabled by default but honestly that makes sense (plus it removes an option which is always nice). Thoughts @iuioiua?

Ah I thought that was what you meant, but re-reading your post I see I misunderstood.

**Pros of always using jitter:**

1. Likely to significantly decrease number of failed API calls (especially those due to 429 Too Many Requests) in real-world scenarios
2. Likely to significantly decrease time-to-completion in real-world scenarios, mainly as a result of 1, but also because the average time between calls is spaced closer together
3. API remains completely unchanged

**Pros of having an option for jitter, disabled by default:**

1. Default behavior remains completely unchanged
2. More deterministic... sort of? I'm not sure whether an unknown power of `m` between `x` and `y` is meaningfully more deterministic than a completely random number between `x` and `y` here. Is there any reason a consumer would specifically want the number of seconds between each retry to be a power of 2?

**Pros of having an option for jitter, enabled by default:**

1. Gives the most-likely preferred behavior as default but allows opt-in to the old behavior

File: async/retry.ts [hunk 1]
```diff
@@ -12,15 +12,15 @@ export class RetryError extends Error {
 export interface RetryOptions {
   /** How much to backoff after each retry. This is `2` by default. */
   multiplier?: number;
-  /** The maximum milliseconds between retries. This is `60000` by default. */
+  /** The maximum milliseconds between retries. This is `60000` by default. `-1` indicates no maximum. */
```
Comment (kt3k) [inline async/retry.ts:15]: > `-1` indicates no maximum.

Why is this handling necessary? The users can pass `maxTimeout: Infinity` if they don't want maximum timeout

File: async/retry_test.ts [hunk 2]
```diff
@@ -31,11 +31,90 @@ Deno.test("[async] retry fails after max errors is passed", async function () {
...
+      for (let i = 0; i < 10; ++i) {
+        const result = _exponentialBackoffWithJitter(cap, base, i, 2);
+        total += Math.round(result);
+        results.push(Math.round(result));
+      }
+
+      assertEquals(results as unknown, row);
+    }
+
+    assertEquals(total, expectedTimingData.total);
```
Comment (kt3k) [inline async/retry_test.ts:116]: Asserting total value feels redundant to me

File: async/retry.ts (hunk 1 above)
Comment (lionel-rowe) [reply async/retry.ts:15]: I just added that to document the existing behavior as of the current version. I guess removing the behavior wouldn't be a breaking change as it's currently undocumented.

File: async/retry_test.ts (hunk 2 above)
Comment (lionel-rowe) [reply async/retry_test.ts]: Removed

**Review by kt3k — APPROVED** [review]
LGTM. Nice!
